### PR TITLE
Add test to verify available memory with load

### PR DESCRIPTION
**Pull request by SuvarnaMeenakshi**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
Added a new test to check AvailableMemory obtained from memUsage OID: 1.3.6.1.4.1.6027.3.10.1.2.9.1.5.1 (implemented in sonic_ax_impl).
Observing that the Available memory computed based on memory usage OID "1.3.6.1.4.1.6027.3.10.1.2.9.1.5.1" is closer to the Available memory of the host. Loading of memory is also done on the host.

Details on why this test is required is added here: https://github.com/sonic-net/sonic-mgmt/issues/8935

#### How did you do it?
memUsage OID: 1.3.6.1.4.1.6027.3.10.1.2.9.1.5.1 (implemented in sonic_ax_impl).
Available memory = ((100-memUsage OID)/100)*memTotal

#### How did you verify/test it?
Ran the new test multiple times on different platforms and verified it is passing.

### CODE REVIEW

File: tests/snmp/memory.py [hunk 1]
```diff
@@ -1,3 +1,11 @@
...
-load = [' ' * 512000000]
+# memory.py <total_memory>
+import sys
+length = 512000000
+iterations = 50
+if len(sys.argv) > 1:
+    total_memory = sys.argv[1]
+    if total_memory <= 4 * 1024 * 1024:
+        length = 320000000
+load = [' ' * length]
```
Comment (lipxu) [inline tests/snmp/memory.py:10]: The length is based on the total memory. If the free memory is not enough, for example, less than 320000000, would it cause memory exhaustion by "load = [' ' * length]" and system crash?

File: tests/snmp/test_snmp_memory.py [hunk 2]
```diff
@@ -142,6 +144,31 @@ def test_snmp_memory_load(duthosts, enum_rand_one_per_hwsku_hostname, localhost,
                   "sysTotalFreeMemory differs by more than {}".format(percentage))
 
 
+def test_snmp_avail_memory_load(duthosts, enum_rand_one_per_hwsku_hostname, localhost, creds_all_duts, load_memory):
```
Comment (lipxu) [inline tests/snmp/test_snmp_memory.py:147]: Sorry, one basic question, just want to know why is the "load_memory" function necessary here? thanks

File: tests/snmp/test_snmp_memory.py [hunk 3]
```diff
@@ -142,6 +144,31 @@ def test_snmp_memory_load(duthosts, enum_rand_one_per_hwsku_hostname, localhost,
...
+    percentage = get_percentage_threshold(int(mem_total))
+
+    mem_avail = int(duthost.shell("grep MemAvailable /proc/meminfo | awk '{print $2}'")['stdout'])
+    snmp_memory_usage_pc = int(localhost.shell(snmp_command)['stdout'])
+    snmp_avail_memory = ((100 - snmp_memory_usage_pc)*mem_total)/100
+    logger.info("SNMP Available Memory: {}".format(snmp_avail_memory))
+    logger.info("DUT Available Memory: {}".format(mem_avail))
+    logger.info("Difference: {}".format(CALC_DIFF(int(snmp_avail_memory), mem_avail)))
+    pytest_assert(CALC_DIFF(int(snmp_avail_memory), mem_avail) < percentage,
+                  "SNMP available memory differs by more than {}".format(percentage))
```
Comment (lipxu) [inline tests/snmp/test_snmp_memory.py:169]: Can we also include the current difference value in the assert log? then we can check the gap in the test result, no need to find the difference value in the case log, thanks.

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml [hunk 4]
```diff
@@ -975,6 +975,12 @@ show_techsupport/test_techsupport.py::test_techsupport:
 #######################################
 #####            snmp             #####
 #######################################
+snmp/test_snmp_memory.py::test_snmp_memory_load:
+  skip:
+    resason: "Test is flaky on some of the platforms"
+    conditions:
+      - https://github.com/sonic-net/sonic-mgmt/issues/8935
```
Comment (lipxu) [inline tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:982]: Just want to double-confirm, test_snmp_memory_load case would be replaced with test_snmp_avail_memory_load, or we still need it, but need to do further analyse before remove skip, thanks.

Comment (SuvarnaMeenakshi) [reply tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:982]: yes, that is right, further analysis is required to check if we can remove the test.
It seems like that test is not correct.
I have updated my findings in the issue. We increase memory load on the host and snmp free memory is obtained from SNMP docker which is not matching the host free memory.

File: tests/snmp/memory.py (hunk 1 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/memory.py:10]: yes, that can happen, from my testing, I have not observed that happening.

File: tests/snmp/test_snmp_memory.py (hunk 2 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_memory.py:147]: load_memory is starting a python script which prints the empty character string set to a specific length.
The memory usage increases when this script is running

File: tests/snmp/memory.py [hunk 5]
```diff
@@ -1,3 +1,11 @@
 #!/usr/bin/python
-load = [' ' * 512000000]
+# memory.py <total_memory>
+import sys
+length = 512000000
+iterations = 50
```
Comment (qiluo-msft) [inline tests/snmp/memory.py:5]: >iterations [](http://example.com/codeflow?start=0&length=10)

Unused?

**Comment by qiluo-msft** [comment]
@AndriiLozovyi Could you help review?

File: tests/snmp/test_snmp_memory.py (hunk 3 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/test_snmp_memory.py]: Add difference value in log.

File: tests/snmp/memory.py (hunk 5 above)
Comment (SuvarnaMeenakshi) [reply tests/snmp/memory.py]: Yes, this is not used, removed.

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml [hunk 6]
```diff
@@ -975,6 +975,12 @@ show_techsupport/test_techsupport.py::test_techsupport:
 #######################################
 #####            snmp             #####
 #######################################
+snmp/test_snmp_memory.py::test_snmp_memory_load:
+  skip:
```
Comment (qiluo-msft) [inline tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:979]: >skip [](http://example.com/codeflow?start=2&length=4)

The purpose of this PR is to skip a testcase based on github issue. Let's only do one thing.

If you change test case at the same time, it is not tested in PR checker.

Comment (SuvarnaMeenakshi) [reply tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:979]: we are skipping test_snmp_memory_load - this check for snmp free memory OID.

I have added a new test test_snmp_avail_memory_load - this is to check the available memory based on memUsage OID.
This test will also increase memory load using the same load_memory() method

Comment (SuvarnaMeenakshi) [reply tests/common/plugins/conditional_mark/tests_mark_conditions.yaml]: Split the diff, created a separate PR to skip test_snmp_memoy_load test: https://github.com/sonic-net/sonic-mgmt/pull/9034

**Comment by abdosi** [comment]
PR needs update based on similar changes on master.

**Comment by SuvarnaMeenakshi** [comment]
Closing this as https://github.com/sonic-net/sonic-mgmt/pull/9113/files was merged to fix the snmp_memory_load test

**Pull request closed without merging**